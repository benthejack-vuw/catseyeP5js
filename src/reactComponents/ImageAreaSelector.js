import React from 'react'

export default class ImageAreaSelector extends React.Component{

	constructor(props){
        super(props);

        this.state = {
            size:props.size,
            coords: [{x:0, y:0}, {x:props.size.x, y:0}, {x:props.size.x, y:props.size.y}],
            selected: undefined
        }

        document.addEventListener('mouseup', this.stopDragging);
    }

    componentDidMount = () => {
        this.centerImage();
        this.constrainSelectorsToImage();
    }

    componentWillUnmount = () => {
        document.removeEventListener('mouseup', this.stopDragging);
    }

    updateMouseCoords = (e) => {
        const rect = document.getElementById(`${this.idPrefix()}-container`).getBoundingClientRect();
        const x = e.clientX - rect.left; 
        const y = e.clientY - rect.top;

        this.setState((prev)=>{
            const coords = [...prev.coords];
            if(prev.selected !== undefined){
                coords[prev.selected] = {x,y};
            }
            return {coords};
        });

        this.constrainSelectorsToImage();
        this.props.updateCallback(this.uvCoordinates());
    }

    setFromTextureCoordinates = (textureCoordinates) => {
        const bounds = this.imageBounds();

        const coords = textureCoordinates.map((coord)=>{
            return {
                x:coord.x*bounds.width + bounds.left,
                y:coord.y*bounds.height + bounds.top
            }
        });

        this.setState(()=>({ coords }))
    }

    uvCoordinates = () => {
        const bounds = this.imageBounds();

        return this.state.coords.map((coord)=>{
            return {
                x:(coord.x - bounds.left)/bounds.width,
                y:(coord.y - bounds.top)/bounds.height
            }
        });
    }
    
    imageBounds = () => {
        const containerRect = document.getElementById(`${this.idPrefix()}-container`).getBoundingClientRect();
        const imageRect = document.getElementById(`${this.idPrefix()}-selecton-image`).getBoundingClientRect();
        return new DOMRect(imageRect.x - containerRect.x, imageRect.y - containerRect.y, imageRect.width, imageRect.height);
    }

    centerImage = () => {
        const imageElem = document.getElementById(`${this.idPrefix()}-selecton-image`);
        const imageRect = imageElem.getBoundingClientRect();
        imageElem.setAttribute('x', (this.state.size.x-imageRect.width)/2 );
        imageElem.setAttribute('y', (this.state.size.y-imageRect.height)/2 );
    }

    constrainSelectorsToImage = () => {
        const bounds = this.imageBounds();
        this.setState(({coords}) => ({
            coords: coords.map(({x, y}) => ({
                x: Math.min(Math.max(bounds.left, x), bounds.right),
                y: Math.min(Math.max(bounds.top, y), bounds.bottom),
            }))
        }))
    }

    idPrefix = () => this.props.id || 'image-area-selector'

    startDragging = (index) => {
        this.setState( ({selected:prevSelected}) => ({selected: !prevSelected ? index : prevSelected}) ) 
    }

    stopDragging = () => {
        this.setState( () => ({selected: undefined}) )
    }

    dashedLines = () => {
        return this.state.coords.map((pt, index, array) => (
            <line 
                key={`selectionLine${index}`}
                strokeDasharray="5, 5"
                x1={pt.x}
                y1={pt.y}
                x2={array[(index+1)%array.length].x}
                y2={array[(index+1)%array.length].y}
                stroke='black'
            />
        ))
    }

    selectionCircles = () => {
        return this.state.coords.map((pt, index) => (
            <circle 
                key={`selection${index}`}
                className='image-selection-circle'                            
                cx={pt.x}
                cy={pt.y}
                r={20}
                onMouseDown = {()=>{this.startDragging(index)}}
            />
        ))
    }

    render = () => ( 
        <div>
            <svg id={`${this.idPrefix()}-container`}
             width={this.state.size.x}
             height={this.state.size.y}
             onMouseMove={this.updateMouseCoords}
            >
                <image 
                    id={`${this.idPrefix()}-selecton-image`}
                    xlinkHref={this.props.image}
                    width='100%'
                    onLoad={
                        (() => {
                            this.centerImage()
                            this.constrainSelectorsToImage()
                            this.setFromTextureCoordinates(this.props.textureCoordinates)
                        })
                    }
                />

                {this.dashedLines()}
                {this.selectionCircles()}
            </svg>
        </div>
    )

}
