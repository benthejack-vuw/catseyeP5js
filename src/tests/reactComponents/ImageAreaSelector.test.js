import React from 'react';
import { shallow } from 'enzyme';
import ImageAreaSelector from '../../reactComponents/ImageAreaSelector';
import { defaults } from '../fixtures/PatternBuilderSettings';

const panelSize = 400;

window.document.getElementById = ()=>({
    getBoundingClientRect: ()=>({x:0, y:0, left:0, right:panelSize, top:0, bottom:panelSize, width:panelSize, height:panelSize}),
    setAttribute: ()=>{}
})

window.DOMRect = (x, y, w, h)=>({x:x, y:y, left:x, right:x+w, top:y, bottom:y+h, width:w, height:h})

test('should render ImageAreaSelector correctly', ()=>{
    const wrapper = shallow(<ImageAreaSelector
                            image={defaults.image}
                            size={{x:panelSize,y:panelSize}}
                            updateCallback={()=>{}}
                            onImageLoad={()=>{}}
                        />);
    expect(wrapper).toMatchSnapshot();
})

test('should call updateCallback prop with current texture coordinates (on mouse moved)', ()=>{
    const mockUpdate = jest.fn(); 
    const wrapper = shallow(<ImageAreaSelector
        image={defaults.image}
        size={{x:panelSize,y:panelSize}}
        updateCallback={mockUpdate}
        onImageLoad={()=>{}}
    />);
    wrapper.find('svg').simulate('mousemove', {clientX: 0, clientY:0});
    expect(mockUpdate).toHaveBeenLastCalledWith([{x:0, y:0}, {x:1, y:0}, {x:1, y:1}]);
})

test('should select node on mouse down', ()=>{

    const wrapper = shallow(<ImageAreaSelector
        image={defaults.image}
        size={{x:panelSize,y:panelSize}}
        updateCallback={()=>{}}
        onImageLoad={()=>{}}
    />);

    wrapper.find('circle').at(1).simulate('mousedown', {});
    expect(wrapper.state('selected')).toBe(1);
})

test('should update mouse coordinates on selected node when mousemoved', ()=>{
    const state = {
            size:{x:panelSize,y:panelSize},
            coords: [{x:0, y:0}, {x:panelSize, y:0}, {x:panelSize, y:panelSize}],
            selected: 1,
    }

    const wrapper = shallow(<ImageAreaSelector
        image={defaults.image}
        size={{x:panelSize,y:panelSize}}
        updateCallback={()=>{}}
        onImageLoad={()=>{}}
    />);

    wrapper.setState(()=>state)
    wrapper.find('svg').simulate('mousemove', {clientX: 123, clientY:321});
    expect(wrapper.state('coords')).toEqual([{x:0, y:0}, {x:123, y:321}, {x:panelSize, y:panelSize}]);
})

test('should NOT update mouse coordinates if no Node selected when mousemoved', ()=>{
    
    const coords = [{x:0, y:0}, {x:panelSize, y:0}, {x:panelSize, y:panelSize}];

    const state = {
            size:{x:panelSize,y:panelSize},
            coords,
            selected: undefined,
    }

    const wrapper = shallow(<ImageAreaSelector
        image={defaults.image}
        size={{x:panelSize,y:panelSize}}
        updateCallback={()=>{}}
        onImageLoad={()=>{}}
    />);

    wrapper.setState(()=>state)
    wrapper.find('svg').simulate('mousemove', {clientX: 123, clientY:321});
    expect(wrapper.state('coords')).toEqual(coords);
})

test('should constrain selectors to image size', ()=>{

    const state = {
            size:{x:panelSize,y:panelSize},
            coords: [{x:0, y:0}, {x:panelSize, y:0}, {x:panelSize, y:panelSize}],
            selected: 1,
    }

    const wrapper = shallow(<ImageAreaSelector
        image={defaults.image}
        size={{x:panelSize,y:panelSize}}
        updateCallback={()=>{}}
        onImageLoad={()=>{}}
    />);

    wrapper.setState(()=>state)

    wrapper.find('svg').simulate('mousemove', {clientX: 1000, clientY:1000})
    expect(wrapper.state('coords')).toEqual([{x:0, y:0}, {x:panelSize, y:panelSize}, {x:panelSize, y:panelSize}]);

    wrapper.find('svg').simulate('mousemove', {clientX: -1000, clientY:-1000})
    expect(wrapper.state('coords')).toEqual([{x:0, y:0}, {x:0, y:0}, {x:panelSize, y:panelSize}]);
})

test('should correctly set node/handle coordinates from normalized texture coordinates', ()=>{
    const texCoords = [{x:0.5, y:0.75}, {x:0.25, y:0.15}, {x:0.6, y:0.85}]
    
    const wrapper = shallow(
        <ImageAreaSelector
            textureCoordinates={texCoords}
            image={defaults.image}
            size={{x:panelSize,y:panelSize}}
            updateCallback={()=>{}}
        />
    );

    wrapper.find('image').simulate('load', {});
    expect(wrapper.state('coords')).toEqual(texCoords.map((coord)=>({x:coord.x*panelSize, y:coord.y*panelSize})));
})