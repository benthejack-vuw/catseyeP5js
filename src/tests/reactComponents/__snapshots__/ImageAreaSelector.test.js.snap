// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ImageAreaSelector correctly 1`] = `
<div>
  <svg
    height={400}
    id="image-area-selector-container"
    onMouseMove={[Function]}
    width={400}
  >
    <image
      id="image-area-selector-selecton-image"
      onLoad={[Function]}
      width="100%"
      xlinkHref="data:image/gif;base64,R0lGODlhEAAQAMQAAORHHOVSKudfOulrSOp3WOyDZu6QdvCchPGolfO0o/XBs/fNwfjZ0frl3/zy7////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAkAABAALAAAAAAQABAAAAVVICSOZGlCQAosJ6mu7fiyZeKqNKToQGDsM8hBADgUXoGAiqhSvp5QAnQKGIgUhwFUYLCVDFCrKUE1lBavAViFIDlTImbKC5Gm2hB0SlBCBMQiB0UjIQA7"
    />
    <line
      key="selectionLine0"
      stroke="black"
      strokeDasharray="5, 5"
      x1={0}
      x2={400}
      y1={0}
      y2={0}
    />
    <line
      key="selectionLine1"
      stroke="black"
      strokeDasharray="5, 5"
      x1={400}
      x2={400}
      y1={0}
      y2={400}
    />
    <line
      key="selectionLine2"
      stroke="black"
      strokeDasharray="5, 5"
      x1={400}
      x2={0}
      y1={400}
      y2={0}
    />
    <circle
      className="image-selection-circle"
      cx={0}
      cy={0}
      key="selection0"
      onMouseDown={[Function]}
      r={20}
    />
    <circle
      className="image-selection-circle"
      cx={400}
      cy={0}
      key="selection1"
      onMouseDown={[Function]}
      r={20}
    />
    <circle
      className="image-selection-circle"
      cx={400}
      cy={400}
      key="selection2"
      onMouseDown={[Function]}
      r={20}
    />
  </svg>
</div>
`;
